File: python/algorithm/cracking/9.9.py
# -*- coding:utf-8 -*-
class Queens:
    def dfs(self, n, cols, xy_diff, xy_sum):
        x = len(cols)
        if x == n:
            self.rst += 1
        else:
            for y in range(n):
                if y not in cols and (x-y) not in xy_diff and (x+y) not in xy_sum:
                    cols.add(y)
                    xy_diff.add(x-y)
                    xy_sum.add(x+y)
                    self.dfs(n, cols, xy_diff, xy_sum)
                    cols.remove(y)
                    xy_diff.remove(x-y)
                    xy_sum.remove(x+y)

    def nQueens(self, n):
        self.rst = 0
        self.dfs(n, set(), set(), set())
        return self.rst



# -*- coding:utf-8 -*-
class Queens2:
    def nQueens(self, n):
        # write code here
        self.rst = 0
        self.util(n, [0]*n, 0)
        return self.rst

    def util(self, n, cols, x):
        if x == n:
            self.rst += 1
        else:
            for y in range(n):
                cols[x] = y
                if self.place(cols, x):
                    self.util(n, cols, x+1)

    def place(self, cols, x):
        for i in range(x):
            if cols[i] == cols[x] or cols[i]-cols[x] == i-x or cols[i]+i == cols[x]+x:
                return False
        return True


if __name__ == '__main__':
    from testfunc import test
    test_data = [
        (
            1,
            1
        ),
        (
            2,
            0
        ),
        (
            3,
            0
        ),
        (
            12,
            14200
        ),
    ]
    # test(Queens().nQueens, test_data)
    test(Queens2().nQueens, test_data)

File: python/algorithm/cracking/9.9.cpp
#include <iostream>
#include <set>

using namespace std;


// very slow, ~3s
class Queens {
public:
    int nQueens(int n) {
        // write code here
        int rst = 0;
        set<int> cols = set<int>();
        set<int> xy_diff = set<int>();
        set<int> xy_sum = set<int>();
        dfs(n, cols, xy_diff, xy_sum, rst);
        return rst;
    }
private:
    void dfs(int n, set<int> &cols, set<int> &xy_diff, set<int> &xy_sum, int & rst)
    {
        int x = cols.size();
        if(x==n)
        {
            ++rst;
        }
        else
        {
            for (int y=0;y<n;++y)
            {
                if(cols.find(y)==cols.end() && xy_diff.find(x-y)==xy_diff.end() && xy_sum.find(x+y)==xy_sum.end())
                {
                    cols.insert(y);
                    xy_diff.insert(x-y);
                    xy_sum.insert(x+y);
                    dfs(n, cols, xy_diff, xy_sum, rst);
                    cols.erase(y);
                    xy_diff.erase(x-y);
                    xy_sum.erase(x+y);
                }
            }
        }
    }
};


// very fast, ~0.3s
class Queens2 {
public:
    int nQueens(int n) {
        // write code here
        int count=0;
        int *a=new int[n+1];
        Queens1(a,1,n,count);
        return count;
         
    }
    void Queens1(int a[],int i,int n,int &count){
        if(i>n){
            count++;
            return ;
        }
        for(int j=1;j<=n;j++){
            a[i]=j;
            if(Place(a,i))
                Queens1(a,i+1,n,count);
        }
    }
    bool Place(int *a,int i){
        for(int j=1;j<i;j++)
            if((a[j]==a[i])||(a[j]-a[i]==(j-i))||(a[j]-a[i]==i-j))
            return 0;
       return 1;
    }
};

int main()
{
    int n;
    cin >> n;
    Queens2 Q = Queens2();
    cout << Q.nQueens(n) << endl;
    return 0;
}
